### PR TITLE
initial checkin for SCCM current branch Azure templates

### PR DESCRIPTION
**Pull request by yizhongwu**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use a parameter for resource locations with the defaultValue set to resourceGroup().location
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values)
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

*
*
*

### CODE REVIEW

File: sccm-currentbranch/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+      "metadata": {
+        "description": "Existing subnet that contains the domain controller."
+      }
+    },
+    "_artifactsLocation": {
+      "type": "string",
+      "metadata": {
+          "description": "The base URI where artifacts required by this template are located including a trailing '/'"
+      },
+      "defaultValue": "https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/sccm-currentbranch"
```
Comment (bmoore-msft) [inline sccm-currentbranch/azuredeploy.json]: add trailing / per description

File: sccm-currentbranch/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+      "type": "string",
+      "defaultValue": "[resourceGroup().location]",
+      "metadata": {
+        "description": "Location for all resources."
+      }
+    }
+  },
+  "variables": {
+    "Quotes": "\"",
+    "storageAccountName": "[concat(toLower(parameters('prefix')), 'sa')]",
```
Comment (bmoore-msft) [inline sccm-currentbranch/azuredeploy.json]: this needs to be globally unique

File: sccm-currentbranch/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+    "containerName": "[concat(parameters('prefix'), 'container')]",
+    "vnetId": "[resourceId(parameters('existingVnetResourcegroup'), 'Microsoft.Network/virtualNetworks', parameters('exsitingVnetName'))]",
+    "generalSettings": {
+      "storageAccountType": "Standard_LRS",
+      "diskstorageAccountType": "Premium_LRS",
+      "virtualMachinessize": "Standard_D2s_v3"
+    },
+    "networkSettings": {
+      "subnetRef": "[concat(variables('vnetId'), '/subnets/',parameters('existingSubnetName'))]",
```
Comment (bmoore-msft) [inline sccm-currentbranch/azuredeploy.json]: simplify, just use resourceId()

File: sccm-currentbranch/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+        "[resourceId('Microsoft.Storage/storageAccounts/', variables('storageAccountName'))]"
+      ],
+      "properties": {
+        "type": "CustomScriptExtension",
+        "publisher": "Microsoft.Compute",
+        "typeHandlerVersion": "1.4",
+        "autoUpgradeMinorVersion": true,
+        "settings": {
+          "fileUris": [
+            "[concat(parameters('_artifactsLocation'),'/scripts/WorkFlow-',variables('vmrole')[copyIndex()],'.ps1',parameters('_artifactsLocationSasToken'))]"
```
Comment (bmoore-msft) [inline sccm-currentbranch/azuredeploy.json]: remove leading / from /scripts and use the uri function

File: sccm-currentbranch/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+        "[resourceId('Microsoft.Storage/storageAccounts/', variables('storageAccountName'))]"
+      ],
+      "properties": {
+        "type": "CustomScriptExtension",
+        "publisher": "Microsoft.Compute",
+        "typeHandlerVersion": "1.4",
+        "autoUpgradeMinorVersion": true,
+        "settings": {
+          "fileUris": [
+            "[concat(parameters('_artifactsLocation'),'/scripts/WorkFlow-',variables('vmrole')[copyIndex(3)],'.ps1',parameters('_artifactsLocationSasToken'))]"
```
Comment (bmoore-msft) [inline sccm-currentbranch/azuredeploy.json]: remove / from /scripts and use the uri function

File: sccm-currentbranch/prereqs/prereq.azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+    "generalSettings": {
+      "domainName": "contoso.com",
+      "storageAccountType": "Premium_LRS",
+      "virtualMachinessize": "Standard_D2s_v3"
+    },
+    "networkSettings": {
+      "virtualNetworkAddressPrefix": "10.0.0.0/16",
+      "subnetAddressPrefix": "10.0.0.0/24",
+      "virtualMachinesIPAddress": "10.0.0.4",
+      "subnetRef": "[concat(variables('vnetId'), '/subnets/',variables('existingSubnetName'))]",
```
Comment (bmoore-msft) [inline sccm-currentbranch/prereqs/prereq.azuredeploy.json]: simplify, just use resourceId()

File: sccm-currentbranch/prereqs/prereq.azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+      "dependsOn": [
+        "[concat('Microsoft.Compute/virtualMachines/',parameters('prefix'),'DC')]"
+      ],
+      "properties": {
+        "publisher": "Microsoft.Powershell",
+        "type": "DSC",
+        "typeHandlerVersion": "2.21",
+        "autoUpgradeMinorVersion": true,
+        "settings": {
+          "ModulesUrl": "https://raw.githubusercontent.com/yizhongwu/azure-quickstart-templates/master/sccm-currentbranch/dsc/CreateADRootDC.ps1.zip",
```
Comment (bmoore-msft) [inline sccm-currentbranch/prereqs/prereq.azuredeploy.json]: See [BP guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#deployment-artifacts-nested-templates-scripts) for how to handle artifacts

File: sccm-currentbranch/scripts/WorkFlow-DPMP.ps1 [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+		#Download azcopy
+		$url = "http://aka.ms/downloadazcopy"
+		Invoke-WebRequest -Uri $url -OutFile $path
+	}
+
+	#Install azcopy
+	Start-Process msiexec.exe -Wait -ArgumentList "/I $path /quiet"
+}
+
+$url = "https://cmsetoolstorage.blob.core.windows.net/work/InServicing/main.ps1"
```
Comment (bmoore-msft) [inline sccm-currentbranch/scripts/WorkFlow-DPMP.ps1]: See [BP guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#deployment-artifacts-nested-templates-scripts) for how to handle artifacts

File: sccm-currentbranch/scripts/WorkFlow-Other.ps1 [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+		#Download azcopy
+		$url = "http://aka.ms/downloadazcopy"
+		Invoke-WebRequest -Uri $url -OutFile $path
+	}
+
+	#Install azcopy
+	Start-Process msiexec.exe -Wait -ArgumentList "/I $path /quiet"
+}
+
+$url = "https://cmsetoolstorage.blob.core.windows.net/work/InServicing/main.ps1"
```
Comment (bmoore-msft) [inline sccm-currentbranch/scripts/WorkFlow-Other.ps1]: See [BP guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#deployment-artifacts-nested-templates-scripts) for how to handle artifacts

File: sccm-currentbranch/scripts/WorkFlow-PS1.ps1 [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+		#Download azcopy
+		$url = "http://aka.ms/downloadazcopy"
+		Invoke-WebRequest -Uri $url -OutFile $path
+	}
+
+	#Install azcopy
+	Start-Process msiexec.exe -Wait -ArgumentList "/I $path /quiet"
+}
+
+$url = "https://cmsetoolstorage.blob.core.windows.net/work/InServicing/main.ps1"
```
Comment (bmoore-msft) [inline sccm-currentbranch/scripts/WorkFlow-PS1.ps1]: See [BP guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#deployment-artifacts-nested-templates-scripts) for how to handle artifacts

File: sccm-currentbranch/scripts/WorkFlow-SQL.ps1 [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+		#Download azcopy
+		$url = "http://aka.ms/downloadazcopy"
+		Invoke-WebRequest -Uri $url -OutFile $path
+	}
+
+	#Install azcopy
+	Start-Process msiexec.exe -Wait -ArgumentList "/I $path /quiet"
+}
+
+$url = "https://cmsetoolstorage.blob.core.windows.net/work/InServicing/main.ps1"
```
Comment (bmoore-msft) [inline sccm-currentbranch/scripts/WorkFlow-SQL.ps1]: See [BP guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#deployment-artifacts-nested-templates-scripts) for how to handle artifacts

File: sccm-currentbranch/azuredeploy.json (hunk 2 above)
Comment (yizhongwu) [reply sccm-currentbranch/azuredeploy.json]: Now using uniqueString func.

File: sccm-currentbranch/azuredeploy.json (hunk 4 above)
Comment (yizhongwu) [reply sccm-currentbranch/azuredeploy.json]: Also, we put all related scripts in the fileUris list.

File: sccm-currentbranch/azuredeploy.json (hunk 2 above)
Comment (yizhongwu) [reply sccm-currentbranch/azuredeploy.json]: We are using uniqueString func now

File: sccm-currentbranch/azuredeploy.json (hunk 4 above)
Comment (yizhongwu) [reply sccm-currentbranch/azuredeploy.json]: Correct. Also put all related scripts in the fileUris list.

File: sccm-currentbranch/prereqs/prereq.azuredeploy.json (hunk 6 above)
Comment (yizhongwu) [reply sccm-currentbranch/prereqs/prereq.azuredeploy.json]: fixed.

File: sccm-currentbranch/scripts/WorkFlow-DPMP.ps1 (hunk 8 above)
Comment (yizhongwu) [reply sccm-currentbranch/scripts/WorkFlow-DPMP.ps1]: Added all using powershell scripts into the folder scripts/.

File: sccm-currentbranch/prereqs/prereq.azuredeploy.json (hunk 7 above)
Comment (yizhongwu) [reply sccm-currentbranch/prereqs/prereq.azuredeploy.json]: If I use _ArtifcatLocation in prereq template, it reports 404 (file not found) error in Travis CI. From the error log, it looks trying to get the file from https://raw.githubusercontent.com/**Azure**/azure-quickstart-template/.... It isn't trying to find it under my fork. So I changed to the hardcoded path. It passed. So I think I will change it back to _ArtifcatLocation once it merges to the main once.

File: sccm-currentbranch/prereqs/prereq.azuredeploy.json (hunk 6 above)
Comment (bmoore-msft) [reply sccm-currentbranch/prereqs/prereq.azuredeploy.json]: not seeing the change here

File: sccm-currentbranch/prereqs/prereq.azuredeploy.json (hunk 7 above)
Comment (bmoore-msft) [reply sccm-currentbranch/prereqs/prereq.azuredeploy.json]: If you author this according to the BP guide it will work...

**Comment by yizhongwu** [comment]
I messed up with this pull request. I will create new branch and new pull request for this.

**Pull request closed without merging**